compiler/optimizer/VPHandlers.cpp

```diff
@@ -4883,13 +4881,19 @@ TR::Node *constrainArraylength(OMR::ValuePropagation *vp, TR::Node *node)
             uintptr_t array = knot->getPointer(kobj->getIndex());
             if (vp->comp()->fej9()->isClassArray(vp->comp()->fej9()->getObjectClass(array)))
                {
-               uintptr_t length = vp->comp()->fej9()->getArrayLengthInElements(array);
-               vp->replaceByConstant(node, TR::VPIntConst::create(vp, length), isGlobal);
-               return node;
+               isKnownObj = true;
+               lowerBoundLimit = vp->comp()->fej9()->getArrayLengthInElements(array);
+               upperBoundLimit = lowerBoundLimit;
                }
             }
          }
+      }
 #endif
+
+   if (isKnownObj && lowerBoundLimit == upperBoundLimit)
```

Comment:
Maybe looking only at the diff does'nt make this obvious but why do we need to check `isKnownObj` here ? i.e. is'nt the other condition sufficient ?

Reply:
[Prior to these changes](https://github.com/eclipse/omr/blob/24984c0b822f942825afb1344130a4a844c23b28/compiler/optimizer/VPHandlers.cpp#L4865-L4892), `constrainArraylength` would first get `arraylength` bounds and `elementSize`, if available, and then check whether the array is a known object of an array class.  If so, it would immediately replace the `arraylength` with the length and return.

The unknown object case where the bound limits are equal is handled later in both [the old version](https://github.com/eclipse/omr/blob/24984c0b822f942825afb1344130a4a844c23b28/compiler/optimizer/VPHandlers.cpp#L4919-L4940) and this version.

The make this commit equivalent to old code, I think it would suffice to change this condition to the following:

```
if (isKnownObj)
```

I was just being a little paranoid by adding the additional check of `lowerBoundLimit == upperBoundLimit`, in case it might be possible for a downstream project to have a known object with an unknown array length, but perhaps that makes no sense.

Reply:
Changed this to check `isKnownObj` alone in commit 73132f854de0ad31d0b20b452875abdd14f31919